### PR TITLE
Allow compiling without QtWebEngine & remove webflow v1

### PR DESCRIPTION
**Pull request by sroracle**
Still enabled by default

Fixes: #856 #932
Supersedes: #1808 #2204
Signed-off-by: Max Rees <maxcrees@me.com>

----

Based on the feedback from #2204 I reworked things so that Flow2 auth is still usable. Indeed it does not require QtWebEngine at all; only Flow1 does. I think the areas I'm not sure about here are the two places where I fall back on using OAuth - should those be Flow2, OAuth, or Basic?

I wrote this based on the 3.2.1 release then rebased it onto master. I verified that Flow2 auth works from the 3.2.1 version. More testing would be appreciated.

### CODE REVIEW

**Comment by sroracle** [comment]
Also, I'm not sure how to handle [`WebFlowCredentials::askFromUser`](/nextcloud/desktop/tree/master/src/gui/creds/webflowcredentials.cpp#L142). Once that's figured out, the following can be added as well for the sake of hygiene (though not strictly necessary):

```diff
diff --git a/src/gui/wizard/owncloudsetuppage.cpp b/src/gui/wizard/owncloudsetuppage.cpp
index 1d20e092d..b6dafbdc2 100644
--- a/src/gui/wizard/owncloudsetuppage.cpp
+++ b/src/gui/wizard/owncloudsetuppage.cpp
@@ -232,8 +232,10 @@ int OwncloudSetupPage::nextId() const
         return WizardCommon::Page_OAuthCreds;
     case DetermineAuthTypeJob::LoginFlowV2:
         return WizardCommon::Page_Flow2AuthCreds;
+#ifdef WITH_WEBENGINE
     case DetermineAuthTypeJob::WebViewFlow:
         return WizardCommon::Page_WebView;
+#endif
     }
     return WizardCommon::Page_HttpCreds;
 }
diff --git a/src/gui/wizard/owncloudwizardcommon.h b/src/gui/wizard/owncloudwizardcommon.h
index 30b4dfb80..a6908898b 100644
--- a/src/gui/wizard/owncloudwizardcommon.h
+++ b/src/gui/wizard/owncloudwizardcommon.h
@@ -46,7 +46,9 @@ namespace WizardCommon {
         Page_HttpCreds,
         Page_OAuthCreds,
         Page_Flow2AuthCreds,
+#ifdef WITH_WEBENGINE
         Page_WebView,
+#endif
         Page_AdvancedSetup,
         Page_Result
     };
diff --git a/src/libsync/networkjobs.h b/src/libsync/networkjobs.h
index 1b0e212d4..f1adf6743 100644
--- a/src/libsync/networkjobs.h
+++ b/src/libsync/networkjobs.h
@@ -446,7 +446,9 @@ public:
         NoAuthType, // used only before we got a chance to probe the server
         Basic, // also the catch-all fallback for backwards compatibility reasons
         OAuth,
+#ifdef WITH_WEBENGINE
         WebViewFlow,
+#endif
         LoginFlowV2
     };
     Q_ENUM(AuthType)
```

**Comment by sroracle** [comment]
Likewise with [`WelcomePage::setupCreateAccountButton`](/nextcloud/desktop/tree/master/src/gui/wizard/welcomepage.cpp#L95), but I think what can be done there is pop open `https://nextcloud.com/register` using the browser like the webview currently does. That would answer one of my earlier questions then - [`OwncloudSetupPage::slotGotoProviderList`](/nextcloud/desktop/tree/master/src/gui/wizard/owncloudsetuppage.cpp) should then be using Flow2 instead of OAuth there if `WITH_WEBENGINE` is undefined.

Or we could just make `WITH_PROVIDERS` depend on `WITH_WEBENGINE`.

File: CMakeLists.txt [hunk 1]
```diff
@@ -130,6 +130,12 @@ if(TOKEN_AUTH_ONLY)
    add_definitions(-DTOKEN_AUTH_ONLY=1)
 endif()
 
+option(WITH_WEBENGINE "Enable old webflow v1 auth using webengine" ON)
```
Comment (mgallien) [inline CMakeLists.txt:133]: Please rename, what you mean is Web Flow v1 auth enabled ?
This has nothing to do with web engine.

File: CMakeLists.txt [hunk 2]
```diff
@@ -130,6 +130,12 @@ if(TOKEN_AUTH_ONLY)
...
+if(WITH_WEBENGINE)
+   message("Enabling flow1 auth")
+   add_definitions(-DWITH_WEBENGINE=1)
```
Comment (mgallien) [inline CMakeLists.txt:136]: If you want to add definitions, do it by target instead of a global one.

File: src/gui/CMakeLists.txt [hunk 3]
```diff
@@ -255,6 +260,12 @@ endif()
 
 add_library(nextcloudCore STATIC ${final_src})
 
+if(WITH_WEBENGINE)
+  set(link_webengine "Qt5::WebEngineWidgets")
```
Comment (mgallien) [inline src/gui/CMakeLists.txt:264]: no, you can have more than one call to `target_link_libraries` that is much better than adding a new variable that you are forced to set empty when you do not need it

File: src/gui/creds/webflowcredentials.cpp [hunk 4]
```diff
@@ -16,7 +16,9 @@
 #include "account.h"
 #include "configfile.h"
 #include "theme.h"
+#ifdef WITH_WEBENGINE
```
Comment (mgallien) [inline src/gui/creds/webflowcredentials.cpp:19]: such compilation condition should be added to `config.h.in`

File: src/gui/creds/webflowcredentialsdialog.cpp [hunk 5]
```diff
@@ -48,12 +52,15 @@ WebFlowCredentialsDialog::WebFlowCredentialsDialog(Account *account, bool useFlo
         connect(this, &WebFlowCredentialsDialog::onActivate, _flow2AuthWidget, &Flow2AuthWidget::slotPollNow);
 
         _flow2AuthWidget->startAuth(account);
-    } else {
+    }
```
Comment (mgallien) [inline src/gui/creds/webflowcredentialsdialog.cpp:55]: else should be on the same line as the previous closing `}`

File: src/gui/wizard/owncloudsetuppage.cpp [hunk 6]
```diff
@@ -127,7 +127,11 @@ void OwncloudSetupPage::slotLogin()
 void OwncloudSetupPage::slotGotoProviderList()
 {
     _ocWizard->setRegistration(true);
+#ifdef WITH_WEBENGINE
```
Comment (mgallien) [inline src/gui/wizard/owncloudsetuppage.cpp:130]: I am pretty sure this is wrong here as you are forcing OAuth instead of Web Flow V2

**Review by mgallien — CHANGES_REQUESTED** [review]

**Comment by sroracle** [comment]
V2: incorporate feedback from @mgallien. Additionally, since the preferred_providers app currently only supports webflow v1 auth to my knowledge, make that option for the client (`WITH_PROVIDERS`) depend on webflow v1 being enabled. This prevents what would be a non-functional "Create account with Provider" button from being displayed.

**Comment by sroracle** [comment]
V3: add missing includes for `config.h`.

**Comment by FlexW** [comment]
@tobiasKaminsky @mgallien should we really bother with introducing a compile option for flow1? I think it Flow1 is not used anymore. Should we then not drop it completely? This will also make the code easier to read and maintain.

**Comment by FlexW** [comment]
@sroracle After discussing internally: Could you remove the flow1 code entirely? We then only need the web engine for the providers page. When WITH_WEBENGINE is disabled, the providers page should be displayed in the users web browser instead of `QWebEngine`. You can use the function `Utility::openBrowser()` in `src/gui/guiutility.cpp`

**Comment by sroracle** [comment]
Thanks for taking the time to discuss it. I will rework it next week.

**Review by mgallien — CHANGES_REQUESTED** [review]
As explained by @FlexW , please modify the PR to remove WebFlow1 and request again review when you are done ?

**Comment by sroracle** [comment]
V4: incorporate feedback from @FlexW @mgallien.
V5: rebase onto master.

**Comment by sroracle** [comment]
V6: silence error from clang-tidy by removing no-longer-needed usage of DetermineAuthTypeJob.

File: src/gui/creds/webflowcredentials.cpp [hunk 7]
```diff
@@ -140,32 +139,17 @@ void WebFlowCredentials::fetchFromKeychain() {
...
-        bool useFlow2 = (type != DetermineAuthTypeJob::WebViewFlow);
-
-        _askDialog = new WebFlowCredentialsDialog(_account, useFlow2);
-
-        if (!useFlow2) {
-            QUrl url = _account->url();
-            QString path = url.path() + "/index.php/login/flow";
-            url.setPath(path);
-            _askDialog->setUrl(url);
-        }
```
Comment (FlexW) [inline src/gui/creds/webflowcredentials.cpp:157]: Why did you remove the auth job? I think you shouldn't remove it.

**Review by FlexW — CHANGES_REQUESTED** [review]

**Comment by FlexW** [comment]
I will have some more comments. I just couldn't finish it today.

File: src/gui/creds/webflowcredentials.cpp (hunk 7 above)
Comment (sroracle) [reply src/gui/creds/webflowcredentials.cpp:157]: As far as I can tell it was dead code since it's only needed to distinguish whether flowv1 is needed over v2. Do you want an error message or something in that case instead?

Comment (FlexW) [reply src/gui/creds/webflowcredentials.cpp:157]: For me, it looks like it also checked if basic authentication is needed.

**Comment by FlexW** [comment]
@sroracle I've realized that it might be not feasible at the moment to remove the flow1 entirely. There were some problems that I wasn't aware of. Sorry for this. To move this forward a little bit quicker I created another pull request that allows compiling without QWebEngine. We need that pull request to make it possible to compile native clients for Apple M1. I would like to close this pull request in favour of the other. I'm sorry that we then will not be able to merge your pull request, but your pull request helped me to get the other pull request implemented quicker. Please continue to contribute to the desktop client. If you need some inspiration for other issues you could solve, just tell me, I will then give you some ;)

New pull request: https://github.com/nextcloud/desktop/pull/3471
